Add support for GitHub's api with tokens

parona-source opened the pull request:
This is to get around rate limits. For authenticated users the hourly rate is 5000 which should be just fine for this use case.

[03:30] <sam_> any idea how I can make pram know my github login
[03:30] <sam_> with cookies or some smth?
[03:30] <sam_> github are rate-limiting anonymous access now

https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens
https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-authenticated-users

parona-source committed:
Add support for GitHub's api with tokens

This is to get around rate limits. For authenticated users the hourly rate is
5000 which should be just fine for this use case.

Signed-off-by: Alfred Wingate <parona@protonmail.com>

mgorny commented on pram:
```diff
@@ -242,12 +250,22 @@ main() {
-				pr=${pr}.patch
+			if [[ -n ${token} ]]; then
+				pr=${pr%%.patch}
+				pr=${pr/github.com/api.github.com\/repos}
```
can't we get away with a single substitution above wget?

mgorny commented on pram:
```diff
@@ -36,6 +36,9 @@ print_help() {
 	echo "                   GitHub repo to use (default: gentoo/gentoo)"
 	echo "  -s, --signoff    Add Signed-off-by to commits (the default)"
 	echo "  -S, --no-signoff Disable adding Signed-off-by to commits"
+	echo "  -t TOKEN, --token TOKEN"
```
I'm not particularly eager to have people passing secrets over argv. I have a `~/.github-token` file, so perhaps we could use that.

parona-source replied:
This is something I do personally as well. Just wasn't sure about the accepted practises elsewhere. So I'm in favour of .github-token

I still think there should be some option to make it use the api path. So maybe --api?

mgorny replied:
Sorry, I don't understand. You mean option to pass the token explicitly? Sure, you can leave it, just note that it's taken by default from the file.

parona-source replied:
I modified the pr to just be y/n option --api which then uses the api urls and uses .github-token with no ability for the user to feed it from  argv.

mgorny commented on pram:
```diff
@@ -286,7 +308,13 @@ main() {
 	if [[ -n ${pr} ]]; then
-		wget -O "${tempdir}/all.patch" "${pr}" || die "Fetching patch failed"
+		if [[ -n ${token} ]]; then
+			wget --header="Authorization: Bearer ${token}" \
```
Similarly, I wish there was a way to pass it without exposing it to argv.

mgorny commented on pram:
```diff
@@ -190,6 +196,11 @@ main() {
 		shift
 	done
 
+	if [[ -n ${api} ]] && ! [[ -r ~/.github-token ]]; then
```
```suggestion
	if [[ ${api} && ! -r ~/.github-token ]]; then
```

mgorny commented on pram:
```diff
@@ -286,7 +297,15 @@ main() {
+			pr=${pr/github.com/api.github.com\/repos}
+			pr=${pr/\/pull\//\/pulls\/}
+			pr=${pr%%.patch}
+			wget --header="Authorization: Bearer $(cat ~/.github-token)" --header="Accept: application/vnd.github.patch" \
```
```suggestion
			wget --header="Authorization: Bearer $(<~/.github-token)" \
			--header="Accept: application/vnd.github.patch" \
```

mgorny commented on pram:
```diff
@@ -21,6 +21,8 @@ print_help() {
 	echo "Options:"
 	echo "  --am-options OPTIONS"
 	echo "                   Pass additional OPTIONS to git am"
+	echo "  --api            Use the GitHub REST API and authenticate with"
```
Ah, that's what you meant. I guess it's fine, though I'd still enable it by default when token is available.

parona-source replied:
Changed it to be default on if a token is present with the ability to disable with --no-api.

mgorny commented on pram:
```diff
@@ -27,6 +27,8 @@ print_help() {
 	echo "  -e EDITOR, --editor EDITOR"
 	echo "                   Override the editor used (default: \${EDITOR})"
+	echo "  --no-api         Don't use the GitHub REST API even if a personal"
+	echo "                   access token is present at ~/.github-token."
```
```suggestion
	echo "                   access token is present at ~/.github-token"
```

mgorny commented on pram:
```diff
@@ -286,7 +296,16 @@ main() {
 
 	if [[ -n ${pr} ]]; then
-		wget -O "${tempdir}/all.patch" "${pr}" || die "Fetching patch failed"
+		if [[ -n ${api} ]]; then
```
```suggestion
		if [[ -n ${api} && ${pr} == *://github.com/*/pull/* ]]; then
```

or something like that. We don't want to change `pull/` to `pulls/` for other forges.

mgorny commented on pram:
```diff
@@ -286,7 +296,16 @@ main() {
 	if [[ -n ${pr} ]]; then
-		wget -O "${tempdir}/all.patch" "${pr}" || die "Fetching patch failed"
+		if [[ -n ${api} && ${pr} == *://github.com/* ]]; then
+			pr=${pr/github.com/api.github.com\/repos}
```
Will this work correctly when using an URL like `https://github.com/org/repo/commit/<hash>.patch`? I don't think it will, so you probably need to restrict the match further.

parona-source replied:
Forgot that was an option. It looks like the API can manage it but it needs similar treatment with s/commit/commits/

https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#get-a-commit

parona-source replied:
Would you want comments with links to the api? Or just a comments explaining this.

parona-source replied:
Just discovered that `https://github.com/org/repo/pull/<pull id>/commit/<hash>.patch` is also a valid option...

Edit: Atleast in this case it should be safe to just change `https://github.com/org/repo/pull/<pull id>/commit/<hash>.patch` -> `https://github.com/org/repo/commit/<hash>.patch`

mgorny replied:
Yeah, we basically support anything that can accept a `.patch` suffix, and actually use them all. API links will probably be better.

mgorny commented on pram:
```diff
@@ -239,6 +249,10 @@ main() {
 	case ${pr} in
+		*://github.com/*/*/pull/*/commits/*)
+			# Simplify for easier use via the api
+			pr=${pr/pull\/*\/commits/commit}
```
Sorry for being picky, but I think this implicitly removes appending `.patch` if it's missing.

parona-source replied:
In this case these urls would never get it the .patch added. Previously only pull and compare urls would have the .patch added implicitly.

I wouldn't be opposed to adding support for this but id also handle it for `*://github.com/*/*/commit/*` in a separate commit.

mgorny replied:
Wouldn't they? At least looking at the code makes me believe the `*://github.com/*/*/pull/*` case below would catch them, and `pr=${to_close}.patch` would add the suffix.

parona-source replied:
Correct. But `*://github.com/*/*/commit/*` not being handled was true so I handled that in the same branch as `*://github.com/*/*/compare/*`.

mgorny commented on pram:
```diff
@@ -286,7 +300,23 @@ main() {
+		if [[ -n ${api} && (
+			${pr} == *://github.com/*/commit/* ||
+			${pr} == *://github.com/*/compare/* ||
+			${pr} == *://github.com/*/pull/* ) ]]; then
```
```suggestion
			${pr} == *://github.com/*/pull/*
		) ]]; then
```

parona-source committed:
Handle github commit urls the same as compare urls

Signed-off-by: Alfred Wingate <parona@protonmail.com>

mgorny commented on pram:
```diff
@@ -239,12 +249,18 @@ main() {
 	case ${pr} in
+		*://github.com/*/*/pull/*/commits/*)
+			# Simplify for easier use via the api
+			to_close=${pr%.patch}
```
I'm sorry but I don't think we ought to be closing the commit :-).

parona-source replied:
Should pram support a use case where someone could do `https://github.com/gentoo/gentoo/pull/42071/commits/70568bfd1c776b568af4b3d4767cac9eb5efa70a` Where it would apply the patch and add a `Part-of: https://github.com/gentoo/gentoo/pull/42071`?

mgorny replied:
Thinking about it, that would make sense — but I'm not going to insist you bother with it.

parona-source committed:
Handle pr specific commit urls

This won't add partof or close to header to the PR as closing the PR
wouldn't be appropriate if you are only picking one commit from it.

Signed-off-by: Alfred Wingate <parona@protonmail.com>

mgorny commented on pram:
```diff
@@ -239,12 +249,20 @@ main() {
+		*://github.com/*/*/pull/*/commits/*)
+			# Simplify for easier use via the api
+			pr=${pr/pull\/*\/commits/commit}
+			if [[ "${pr}" != *.patch ]]; then
+				pr=${pr}.patch
+			fi
```
```suggestion
			pr=${pr%.patch}.patch
```

Sorry it didn't occur to me earlier. This should work, right?

parona-source committed:
Simplify .patch appending

Signed-off-by: Alfred Wingate <parona@protonmail.com>

mgorny reviewed (approved):
Thanks! Is it ready to merge, or do you want to test more?

parona-source commented:
I'm happy with it as is.
